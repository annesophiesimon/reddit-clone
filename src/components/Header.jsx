import reddit from '../assets/images/reddit-logo.png';
import { HiOutlineSearch } from 'react-icons/hi';

const Header = (data) => {
  const { handleSubmit, handleChange, term } = data;

  return (
    <header className="flex bg-white items-center p-2.5 shadow-lg mb-4">
      <div className="w-10">
        <img alt="reddit logo" src={reddit} />
      </div>
      <form
        className="flex justify-between border border-gray-300 p-1 rounded-lg mx-auto w-80"
        onSubmit={handleSubmit}>
        <input
          type="text"
          id="header-search"
          placeholder="Search Reddit"
          name="search"
          aria-label="Search posts"
          value={term}
          onChange={handleChange}
        />
        <button type="submit" aria-label="Search">
          <HiOutlineSearch />
        </button>
      </form>
    </header>
  );
};

export default Header;
