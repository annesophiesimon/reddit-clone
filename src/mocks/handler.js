import { rest } from 'msw';
import { setupServer } from 'msw/lib/node';

import {
  resultMockSubreddit,
  resultMockSewing,
  resultMockCrossfit,
  resultMockBachata,
  resultMockEmptyPost
} from './result';

const API_ROOT = 'https://www.reddit.com';
export const handlers = [
  rest.get(`${API_ROOT}/subreddits.json`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(resultMockSubreddit), ctx.delay(50));
  }),
  rest.get(`${API_ROOT}/r/sewing.json`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(resultMockSewing), ctx.delay(150));
  }),
  rest.get(`${API_ROOT}/r/crossfit.json`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(resultMockCrossfit), ctx.delay(150));
  }),
  rest.get(`${API_ROOT}/search.json`, (req, res, ctx) => {
    const q = req.url.searchParams.get('q');
    if (q === 'bachata') {
      return res(ctx.status(200), ctx.json(resultMockBachata), ctx.delay(150));
    } else {
      return res(ctx.status(200), ctx.json(resultMockEmptyPost), ctx.delay(150));
    }
  })
];

export const server = setupServer(...handlers);

server.listen();
