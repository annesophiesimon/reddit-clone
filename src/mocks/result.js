export const resultMockSubreddit = {
  data: {
    children: [
      { data: { id: 1, display_name_prefixed: 'r/home', display_name: 'Home' } },
      { data: { id: 2, display_name_prefixed: 'r/sewing', display_name: 'Sewing' } },
      { data: { id: 3, display_name_prefixed: 'r/basket', display_name: 'Basket' } },
      { data: { id: 4, display_name_prefixed: 'r/crossfit', display_name: 'Crossfit' } },
      { data: { id: 5, display_name_prefixed: 'r/tech', display_name: 'Tech' } }
    ]
  }
};

export const resultMockSewing = {
  data: {
    children: [
      {
        data: {
          id: 1,
          ups: 26,
          title: 'Mon nouveau manteau cousu main',
          url: 'image.jpg',
          author: 'Kevin',
          created_utc: 36367,
          num_comments: 256
        }
      },
      {
        data: {
          id: 2,
          ups: 33,
          title: 'Comment monter fermeture Ã©claire',
          url: 'image.jpg',
          author: 'Bernard',
          created_utc: 23578899,
          num_comments: 346
        }
      }
    ]
  }
};

export const resultMockCrossfit = {
  data: {
    children: [
      {
        data: {
          id: 1,
          ups: 26,
          title: 'How to improve my muscle up',
          url: 'image.jpg',
          author: 'Katarina',
          created_utc: 36367,
          num_comments: 78
        }
      }
    ]
  }
};

export const resultMockBachata = {
  data: {
    children: [
      {
        data: {
          id: 1,
          ups: 33,
          title: 'Bachata',
          url: 'image.jpg',
          author: 'Lola',
          created_utc: 3636,
          num_comments: 22
        }
      }
    ]
  }
};

export const resultMockEmptyPost = {
  data: {
    children: []
  }
};

export const resultsMock = [
  resultMockSubreddit,
  resultMockSewing,
  resultMockBachata,
  resultMockEmptyPost
];
