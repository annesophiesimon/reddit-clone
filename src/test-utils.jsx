// create utils folder
import React from 'react';
import store from './App/store';
import { Provider } from 'react-redux';

export const renderWithProviders = (component) => {
  <Provider store={store}>{component}</Provider>;
};
