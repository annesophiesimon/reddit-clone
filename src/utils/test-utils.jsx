import React from 'react';
import PropTypes from 'prop-types';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { setupStore } from '../app/store';

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    Wrapper.propTypes = {
      children: PropTypes.string.isRequired
    };
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
