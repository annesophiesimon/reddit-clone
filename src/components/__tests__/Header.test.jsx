import Header from '../Header';
import React from 'react';
import { renderWithProviders } from '../../utils/test-utils';
import { server } from '../../mocks/handler';

// Enable API mocking before tests.
beforeEach(() => server.listen());

// Disable API mocking after the tests are done.
afterAll(() => server.close());

describe('Header', () => {
  it('renders correctly', () => {
    const header = renderWithProviders(<Header />);
    expect(header.container).toMatchSnapshot();
  });
});
