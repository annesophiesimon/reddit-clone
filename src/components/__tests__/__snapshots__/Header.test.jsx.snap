// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
<header
  className="flex bg-white items-center p-2.5 shadow-lg mb-4"
>
  <div
    className="w-10"
  >
    <img
      alt="reddit logo"
      src="reddit-logo.png"
    />
  </div>
  <form
    className="flex justify-between border border-gray-300 p-1 rounded-lg mx-auto w-80"
  >
    <input
      aria-label="Search posts"
      id="header-search"
      name="search"
      placeholder="Search Reddit"
      type="text"
    />
    <button
      aria-label="Search"
      type="submit"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
        />
      </svg>
    </button>
  </form>
</header>
`;
